src/broker/Manager.cc

```diff
@@ -1013,20 +1114,112 @@ void Manager::Process() {
     }
+}
+
+namespace {
 
-    for ( auto& s : data_stores ) {
-        auto num_available = s.second->proxy.mailbox().size();
+// Note: copied from Stmt.cc, might be worth to move to a common place.
+EnumValPtr lookup_enum_val(const char* module_name, const char* name) {
+    const auto& id = zeek::detail::lookup_ID(name, module_name);
+    assert(id);
+    assert(id->IsEnumConst());
+
+    EnumType* et = id->GetType()->AsEnumType();
+
+    int index = et->Lookup(module_name, name);
+    assert(index >= 0);
+
+    return et->GetEnumVal(index);
+}
 
-        if ( num_available > 0 ) {
-            had_input = true;
-            auto responses = s.second->proxy.receive(num_available);
+} // namespace
 
-            for ( auto& r : responses )
-                ProcessStoreResponse(s.second, std::move(r));
+void Manager::ProcessLogEvents() {
+    static auto plval = lookup_enum_val("Broker", "LOG");
+    static auto lpli = id::find_type<RecordType>("Broker::Info");
+    static auto ev_critical = lookup_enum_val("Broker", "CRITICAL_EVENT");
+    static auto ev_error = lookup_enum_val("Broker", "ERROR_EVENT");
+    static auto ev_warning = lookup_enum_val("Broker", "WARNING_EVENT");
+    static auto ev_info = lookup_enum_val("Broker", "INFO_EVENT");
+    static auto ev_verbose = lookup_enum_val("Broker", "VERBOSE_EVENT");
+    static auto ev_debug = lookup_enum_val("Broker", "DEBUG_EVENT");
+
+    auto evType = [](BrokerSeverityLevel lvl) {
+        switch ( lvl ) {
+            case BrokerSeverityLevel::critical: return ev_critical;
+            case BrokerSeverityLevel::error: return ev_error;
+            case BrokerSeverityLevel::warning: return ev_warning;
+            case BrokerSeverityLevel::info: return ev_info;
+            case BrokerSeverityLevel::verbose: return ev_verbose;
+            default: return ev_debug;
+        }
+    };
+
+    constexpr const char* severity_names_tbl[] = {"critical", "error", "warning", "info", "verbose", "debug"};
+
+    auto events = bstate->loggerQueue->Drain();
+    while ( ! events.empty() ) {
+        for ( auto& event : events ) {
+            auto severity = event->severity;
+            if ( bstate->logSeverity >= severity ) {
+                auto record = make_intrusive<RecordVal>(lpli);
+                record->AssignTime(0, run_state::network_time);
+                record->Assign(1, evType(event->severity));
+                auto ev = make_intrusive<StringVal>(event->identifier);
+                record->Assign(2, ev);
+                auto msg = make_intrusive<StringVal>(event->description);
+                record->Assign(4, msg);
+                log_mgr->Write(plval.get(), record.get());
+            }
+            if ( bstate->stderrSeverity >= severity ) {
+                fprintf(stderr, "[BROKER/%s] %s\n", severity_names_tbl[static_cast<int>(severity)],
+                        event->description.c_str());
+            }
         }
+        events = bstate->loggerQueue->Drain();
```

Comment:
This is a bit theoretical, but I think it's preferable to yield here to avoid starvation of the IO loop. E.g. what if the `log_mgr->Write()` reaches out to broker_mgr and in turn logs something, filling the queue.

Particular when we go through the event loop, this should go so that we drain the event queue.